### PR TITLE
Client for java spec tests

### PR DESCRIPTION
**Pull request by wottop**
Includes some foundation for the test conversion.

### CODE REVIEW

**Review by nikosmoum — COMMENTED** [review]
Haven't reviewed this yet, but just a comment based on the jenkins build: the java spec tests shouldn't be running as part of the unit test task (`gradlew test`). I would expect them to be run as a separate gradle task (e.g. `gradlew java-spec` or `gradlew spec`) and not be dependent on the unit tests

**Comment by Ceiu** [comment]
I would also vote for `gradlew spec`

**Review by Ceiu — CHANGES_REQUESTED** [review]
The distractions have completely derailed my train of thought, so I'm leaving this review here for the moment. I'll try to give it the attention it deserves on the next round.

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/PropertiesConfiguration.java
Comment (Januson) [inline spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/PropertiesConfiguration.java:33]: Do we need synchronized config? We will be loading it at the beginning and it should stay immutable for the whole run.

**Review by Januson — CHANGES_REQUESTED** [review]

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SystemProperties.java
Comment (Ceiu) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SystemProperties.java:21]: :+1:

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java
Comment (wottop) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java:38]: I'm happy to go with whichever y'all want to.
@nikosmoum  and @Ceiu ?

File: spec-tests/src/test/java/org/candlepin/spec/SpecTest.java
Comment (wottop) [reply spec-tests/src/test/java/org/candlepin/spec/SpecTest.java:28]: So far it is for the common annotations.

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java
Comment (Ceiu) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java:37]: If we have an intention to be able to run these against any arbitrary server and minimize the amount of data left over from that, then we'll need some kind of cleanup on the order of what the current spec tests do. If that is no longer a concern or design constraint, then we can drop it.

If it's not something that's been discussed, I would lean toward doing cleanup as that's what we already have and it's, at least theoretically, not that hard to deal with.

Comment (nikosmoum) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java:37]: We have gone a bit back and forth when discussing this, but I think we wanted to avoid the overhead from doing granular cleanup, so:
- We make sure we have good randomization utility functions (similar to the `random_string()` in rspec) for creating data with unique attributes and avoid duplication errors.
- When running the tests sequentially, we would be running a SQL TRUNCATE on all tables (with exceptions for tables that hold data that drive business logic, such as cp_consumer_type) between each test.
- When running the tests in parallel, we would not be doing any cleanup.

But all that is for later tasks. For now, we should probably start without any cleanup.

Comment (Ceiu) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java:37]: If we're going to be doing TRUNCATES, we literally cannot run the spec suite against arbitrary deployments, as we'd be destroying any pre-existing data that might be there with it. 

If that's understood and accepted, fine, however it calls into question the value of isolating this to its own project/package completely separate from the main code base. I get that it helps us avoid accidentally violating spec test principles, but ... not much else.

Comment (nikosmoum) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java]: Running the spec tests against arbitrary deployments wasn't something we'd discussed, but it could be useful. I think the most important thing here is to avoid performance overhead from having to track and clean test data. So we could very well just not do any cleanup.

Comment (wottop) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java]: I took out the cleanup for now.

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/ApiClientFactory.java
Comment (nikosmoum) [inline spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/ApiClientFactory.java:32]: checkstyle doesn't seem to be evaluating the `spec-test` package. we should find a way to make sure it does, in order to keep the code clean

**Review by nikosmoum — CHANGES_REQUESTED** [review]
I took a quick look; might do another run on Monday

File: spec-tests/config/checkstyle/HEADER.txt
Comment (nikosmoum) [inline spec-tests/config/checkstyle/HEADER.txt:1]: You don't need to duplicate the checkstyle config. If you rename the checkstyle tasks in spec-test build.gradle to `checkstyleSpec` as I mention in my other comment, the main config is picked up already, so you can delete the `spec-tests/config` folder

**Review by nikosmoum — CHANGES_REQUESTED** [review]

**Comment by nikosmoum** [comment]
Also, for some reason, the unit tests are running the spec-tests again

**Review by nikosmoum — APPROVED** [review]
ack from me, but there's still a couple of comments from @Januson to sort out, and we should also squash this into a single commit before merging

**Review by Januson — CHANGES_REQUESTED** [review]
There is an issue with how spec tests are excluded from unit test run. This way we would have to edit run configurations every time we want to run spec tests from IDE instead of just click->run/debug. Also some of my previous comments are still waiting for response.

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java
Comment (wottop) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/SpecTestFixture.java:32]: I made the properties static and made the load synchronized so that it only happens once.
This should avoid more than a single load.

File: spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/PropertiesConfiguration.java
Comment (wottop) [reply spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/PropertiesConfiguration.java:42]: See the above comment

File: spec-tests/build.gradle
Comment (nikosmoum) [reply spec-tests/build.gradle:33]: This doesn't work, because any test filtering/exclude/include is only applied to the current test task, not the subproject test tasks run after it. I don't see a clear way in gradle docs on how to not run subproject tasks when running the parent task, so I think the current way is the simplest thing that works.

Comment (nikosmoum) [reply spec-tests/build.gradle:33]: since the current solution breaks the ability to run spec tests from the IDE, here's an alternative that does what we want, without breaking running from IDE:
- remove the runSpec workaround from the subproject build.gradle
- add this workaround to the top project build.gradle(by replacing `:run` with `:test`): https://stackoverflow.com/questions/51903863/gradle-disable-automatic-subproject-execution-for-specific-task

**Comment by wottop** [comment]
Is my last change wrong? It tried to run then with the unit tests again.

File: spec-tests/build.gradle
Comment (nikosmoum) [reply spec-tests/build.gradle]: @wottop I am able to reproduce the issue when using our docker container in the same way jenkins does: `./docker/test -m -c "cp-test -uuu"`. Apparently, the `-uuu` is an attempt at running the unit tests 3 times (although it doesn't seem it ever  worked).
Another way to reproduce is to simply run the test command twice, without a clean inbetween: `./gradlew clean test; ./gradlew test`. For whatever reason (probably something to do with the fact that gradle is a daemon with caching mechanisms involved that I am not familiar with, so project properties aren't working as expected?).

I tried the workaround I mentioned before (add this to the top project build.gradle test task (by replacing :run with :test): https://stackoverflow.com/questions/51903863/gradle-disable-automatic-subproject-execution-for-specific-task) and it works without triggering this problem.

**Pull request merged**
